Custom confirmation message

boazsender opened the pull request:
Enables custom form submission success confirmation messages in markdown.

#### I have verified this PR works with

-   [x] Online form submission
-   [x] Offline form submission
-   [x] Saving offline drafts
-   [x] Loading offline drafts
-   [ ] Editing submissions
-   [ ] Form preview
-   [ ] None of the above

#### What else has been done to verify that this works as intended?
This PR is being tested in integration with Kobo by the Kobo team.

#### Why is this the best possible solution? Were any other approaches considered?
We discussed this solution at length. Key aspects of the implementation are:
* Looks for a submitMessage node in the form before rendering a custom message node to ensure backward compatibility with forms that already have a message node but don't expect this behavior.
* Renders markdown
* Visually hides the success dialog title but leaves it available in the DOM for assistive technologies.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

If you opt into this feature with a form, submitters will see a custom submit message.

#### Do we need any specific form for testing your changes? If so, please attach one.

You need a form with a submitMessage node and a message node. I have been testing with [centro](https://github.com/jnm/centro). An example message response is available at [jnm/centro/.../app/controllers/submission.js#L51](https://github.com/jnm/centro/blob/14c135666005ed926ffafd75ee5f0d4593dbe0e2/app/controllers/submission.js#L51).

tiritea commented:
Some related background... this custom confirmation message necessitated being able to add a [custom instance attribute](https://xlsform.org/en/#advanced-use-and-extensibility) in the original form with which to specify (to the backend...) what field in the form to use to populate the desired response. This in turn requires, preferably, being able to specify a usual ${foo} identifier in the associated XLSForm as the value for this custom attribute. However, presently pyxform does not correctly parse and resolve ${foo} strings in custom attribute value to their XML element path; instead it merely copies the string "${foo}" as-is. This is discussed in some depth on the ODK Forum here as a desired fix - or new feature depending on your view - to pyxform: https://forum.getodk.org/t/xlsform-spec-also-resolve-foo-names-in-custom-settings-attributes/48735

There is a open issue against pyxform to address this: https://github.com/XLSForm/pyxform/issues/717

lognaturel commented on packages/enketo-express/public/js/src/module/controller-webform.js:
```diff
@@ -443,9 +444,28 @@ function _submitRecord(survey) {
+                    'text/xml'
+                );
+                const maybeSubmitMessage =
+                    xmlForm.querySelector('submitMessage');
```
> Kobo BE picks this up in the originating POST's submission XML

Great, so this looks like code that belongs in Kobo's backend, not in Enketo.

If it is indeed a requirement to also change the text displayed at the returnUrl, I see two options: 
* have your server serve a page to redirect to for that form. If a submit message is specified, it is displayed on that page and that page's URL is included as the return URL for the form
* read the response message from the OpenRosa response as with the no-redirect case. Change the default redirect page shipped with Enketo so that a message can be injected into it

tiritea replied:
Note, some backends are already returning a `<message>` string in the `<OpenRosaResponse>` body, but it is being completely ignored and Enketo is displaying a boilerplate response to the user anyway. So in order to retain this existing desired (but arguably incorrect?) behavior of Enketo - ie not displaying the `<message>` in a lot of cases - the decision was made to first check whether to form just sent actually had a "kobo:submitMessage" setting in it, and only display the `<message>` body if it did [yes, this is a bit of a hack, but necessary to retain existing behavior of Enketo when talking to backends that return a `<message>` today, but it is being ignored...] But, again, this check should be for the presence of the _custom attribute_, not any hardcoded specific XML element .

lognaturel replied:
Ok, I better see the flow here, I was trying to go too quickly. Now I see this is where the message is alerted and has nothing to do with the return url. 

> yes, this is a bit of a hack

Indeed. At the very least I think it should check for the `kobo` namespace. Right now, unless I've misunderstood something else, the first attribute or element with name `submitMessage` in the default namespace will be a match. You could look at some of the OpenClinica-specific behavior for examples of introducing and isolating behavior for a specific platform.

tiritea replied:
So this 'hack' is undesirable, and is only necessary to - ostensibly - persist Enketo's (arguably) incorrect current behavior of completely ignoring the `<message>` being returned in the OpenRosaResponse, and instead displaying a hardcoded message instead. Ideally, a boilerplate message would _only_ be displayed if the message were, say, empty (if even then), whereas any `<message>` actually provided would be displayed, as intended! That would seem to be entirely the intent of having this optional message element in the OpenRosaResponse body.

So my preference would be to remove this hack, and have Enketo always display - for the post-submission popup - whatever `<message>` in received. And to only display the existing generic boilerplate message when no message is provided.

The potential impact of doing this (ie removing this hack, for checking for the special attribute in the original form...) is that Enketo would now actually start displaying whenever OpenRosaResponse `<message>`s that existing backends might already happen to be returning (!), but which previously Enketo has been hiding. I dont think this is necessarily a 'bad' thing, per se, and if this is undersirable then these backends can always start returning an empty `<message>` instead to retain Enketo's existing boilerplate followup response.

Again, the explicit intent of the `<message>` in the OpenRosaResponse is to provide a mechanism for a BE to return a customized post-submission message to the client upon successful submission. Changing Enketo to allow this to include markdown seems like an obvious, consistent, and pretty natural way to accomplish this desired feature.

lognaturel replied:
> Enketo's (arguably) incorrect current behavior of completely ignoring the <message> being returned in the OpenRosaResponse

I agree that the original spec intent was for the message to come from the server. I think both Collect and Enketo ended up taking control of success messages and some standard error messages because it means they can provide a more uniform, coherent user experience and take control of localization.

I'm not immediately against changing that behavior but I think it warrants its own broader discussion pulling in other server implementers. I think we'd also want to have a plan for localization. That would entail having Enketo send `Accept-Language` headers (https://docs.getodk.org/openrosa-http/#http-requests-and-responses). Servers would then be expected to have localized default messages to return. I think it's a fair amount of work across a lot of code bases.

A somewhat less hacky option would be to document a special message `nature` (https://docs.getodk.org/openrosa-http/#http-responses) that clients could recognize and use to know they should override their default strings. For example, it could have a value like `__client_default_override`. Servers could specify messages of this nature when they know they have something more useful than the client default to say.

lognaturel replied:
Oh, and I don't know whether this is what @boazsender might be referring to with "moving this work to the single route" but you could also entirely bypass OpenRosa messages by targeting the path with `returnUrl` instead.

boazsender replied:
I was confused when I said I was moving this work to the `single` route. Please disregard that comment.

lognaturel commented on packages/enketo-express/public/js/src/module/connection.js:
```diff
@@ -164,30 +164,22 @@ function _uploadBatch(recordBatch) {
-                    const xmlResponse = parser.parseFromString(
-                        text,
-                        'text/xml'
+            return response.text().then((text) => {
```
Addressed at https://github.com/enketo/enketo/pull/1368/files#r1866695475 -- this is reading the message and returning it.

lognaturel commented on packages/enketo-express/public/js/src/module/controller-webform.js:
```diff
@@ -443,9 +444,28 @@ function _submitRecord(survey) {
+                    xmlForm.querySelector('submitMessage');
+                msg = maybeSubmitMessage
+                    ? result.message.length > 0
+                        ? jstransformer.render(result.message)
```
Any XSS opportunities?

boazsender replied:
Great point. I'm looking around enketo for a markdown sanitization utility. Any pointers?

jnm commented:
Thank you, @lognaturel, for your review. We're going to continue revising this PR significantly, so please consider it to be a draft at this time—and @boazsender will mark it as such :)

boazsender commented:
Marked this back to draft. I'm moving this work to the `single` route, and will address your other comments when I do @lognaturel. Thanks for the great review!

boazsender commented on packages/enketo-express/public/js/src/module/controller-webform.js:
```diff
@@ -340,9 +340,9 @@ function _loadRecord(survey, instanceId, confirmed) {
  * and is not used in offline-capable views.
  *
- * @param {Survey} survey
+ * @param {Data} data
```
My bad, thanks for catching this. Switching to ` @param {Survey} survey` and `@param {string} modelStr` in my next commit.

boazsender commented:
I was confused when I said I was moving this work to the `single` route. I was trying to figure out why my implementation wasn't working in the `single` route before I realized that my test form had a redirect url, which this implementation didn't account for. I'm working on an update that supports redirect urls.

I also know we're having a broader conversation about whether this work would be accepted at all. Completely understand your concerns. I'm going to update my branch (and this PR) to include my new implementation so that Kobo can use it. We can continue the conversation about whether this code makes sense here. Just didn't want you to be surprised by the changes to the branch.

boazsender commented:
Pushed a new commit which adds support for custom messages to forms with redirects, and addressed the type integrity feedback (thanks for that @lognaturel). I also rebased from main to keep this PR up to date. I can also address the markdown sanitization feedback with a little more guidance on how that's done in this project.

boazsender commented:
Oh, and I also made some styling changes for custom messages in this latest commit.

boazsender commented:
And one more commit for tidier scroll bars.

boazsender commented:
I'm closing this PR for now based on the discussion above and to improve the proposed functionality based on initial testing. We'll make a new PR once these things have been addressed and to allow for upfront discussions on the forum.

tiritea commented:
Please note, the change highlighted above:

```
                const maybeSubmitMessage =
                    xmlForm.querySelector('submitMessage');

```                    
this will - **_in the absence of any other checks regarding the presence of an actual custom form attribute_** (eg **attribute::kobo:submitMessage**) - cause Enketo to believe that the form may contain a custom submit message and therefore to look out for a `<message>` in the `<OpenRosaResponse>` and change its post-submit behavior accordingly in any case. This is incorrect, and will result in Enketo behaving differently for any form containing any _element_ that happens to be called 'submitMessage'.

For example, this form:

[SimpleFormBad.xlsx](https://github.com/user-attachments/files/19493828/SimpleFormBad.xlsx)

has nothing to do with this custom post-submit message feature, but just happens to have a question named 'submitMessage'. But the behavior of Enketo with this form is now very different if I instead had called the question to 'name'. Again, the specific name of the XML element isnt the trigger for the new behavior; the trigger is the presence of the custom **attribute::kobo:submitMessage** form attribute (and the form having an element with the corresponding specified path, whatever it may be)
